Fix for the node-agent privs and volumes

mpryc opened the pull request:
Depends-On: https://github.com/openshift/velero/pull/361 (this one will get into openshift/velero from upstream via rebase as the upstream got already merged)

Fixes:
-  [OADP-5033](https://issues.redhat.com//browse/OADP-5033)
-  [OADP-5104](https://issues.redhat.com//browse/OADP-5104)
   - With default to be permissive (FSBackup working)
-  [OADP-5105](https://issues.redhat.com//browse/OADP-5105)
   - With default to be permissive (FSBackup working)

Partial fix for:
- [OADP-5031](https://issues.redhat.com//browse/OADP-5031) - need to address doc and sample apps
-  [OADP-5141](https://issues.redhat.com//browse/OADP-5141) - need to address datamover ?
-  [OADP-5182](https://issues.redhat.com//browse/OADP-5182) - need to address datamover ?

# About the change
This changes adds:
 - New configuration option to the `velero` under DPA Spec
 - The `disableFsBackup`  default is **false** or **non-existing**, which is same as **false**
   ```yaml
    spec:
    [...]
      configuration:
        [...]
        velero:
          disableFsBackup: true <----- NEW OPTION
   ```
 - Modifies that the node-agent is `always` ran as non-root user !
 - Changes the root filesystem to be read only !
   - Makes two mountpoints writeable:
     - `/home/velero`
     - `/tmp/credentials`
 - Uses `SeccompProfileTypeRuntimeDefault` for the `SeccompProfile`

 - With the `disableFsBackup` set to **false** or **non-existing**, the following options are added to the `SecurityContext`:
     - `Privileged: true`
     - `AllowPrivilegeEscalation: true` 
  
 - With `disableFsBackup` flag to set to true:
   - Removes the following mounts from the node-agent:
     - `host-pods`
     - `host-plugins`
   
## Why the changes were made

To enhance node-agent security

## How to test the changes made

1. Deploy with this PR and velero built:
DPA:
```yaml
spec:
[...]
  configuration:
    [...]
    velero:
      disableFsBackup: true <----- NEW OPTION
    nodeAgent:
      enable: true
      uploaderType: kopia
  unsupportedOverrides:
    veleroImageFqin: 'quay.io/migi/velero:latest'
```

mpryc commented:
With this we have node-agent with:

```yaml
spec:
  [...]
  securityContext:
    runAsUser: 1000
    runAsGroup: 1000
    runAsNonRoot: true
    fsGroup: 1000
    seccompProfile:
      type: Unconfined
  containers:
      [...]
      securityContext:
        capabilities:
          drop:
            - ALL
        privileged: false
        readOnlyRootFilesystem: true
        allowPrivilegeEscalation: false
```

mpryc commented:
With this setting Backup to actually backup FS causes node-agent to fail and produces PartiallyFailed backu, so I think that is expected behavior ?

```shell
Phase:  PartiallyFailed (run `velero backup logs backup` for more information)


Errors:
  Velero:    message: /pod volume backup failed: error exposing host path for pod volume: error identifying unique volume path on host for volume mongo-data in pod mongo-6866977c96-lbmsm: expected one matching path: /host_pods/c50b8080-a1e5-46fe-861e-0ebba40a086f/volumes/*/pvc-4b91d7bb-6ea3-44fb-96d3-1cd111ffdeab/mount, got 0
```

mpryc commented:
/hold - need to get the velero change first: https://github.com/openshift/velero/pull/361

sseago commented on api/v1alpha1/dataprotectionapplication_types.go:
```diff
@@ -295,6 +295,13 @@ type VeleroConfig struct {
 	DefaultVolumesToFSBackup *bool `json:"defaultVolumesToFSBackup,omitempty"`
+	// DisableFsBackup determines whether the NodeAgent should disable file system backup.
+	// When set to true, the NodeAgent runs in non-privileged mode.
+	// Privileged mode is required for backing up block devices.
```
Could you clarify this part for me. Block devices are only supported with datamover, not fs-backup. datamover doesn't use the node agent pod for data upload/download, so why does the node agent need to be privileged for this use case?

mpryc replied:
Ah this is from previous implementation where I had one flag for all non-root, so will update.

mpryc replied:
fixed.

mpryc commented on config/crd/bases/oadp.openshift.io_dataprotectionapplications.yaml:
my beloved `make bundle` - fixed, thanks !

mpryc commented:
@mpryc note to myself: check the permissions of the fs files after restoring - if the user perms/owner/group are preserved.

mpryc commented:
Checked the following:
 1) Deployed mysql-persistent application with mysql user
 2) Created from the mysql pod file that persisted on PV: `/var/lib/mysql/data/todolist/my_file`
 ```shell
$ vi /var/lib/mysql/data/todolist/my_file
-rw-r--r--. 1 mysql mysql 22 Mar 12 15:35 /var/lib/mysql/data/todolist/my_file
$ chmod 700 /var/lib/mysql/data/todolist/my_file 
$ ls -lah /var/lib/mysql/data/todolist/my_file
-rwx------. 1 mysql mysql 22 Mar 12 15:36 /var/lib/mysql/data/todolist/my_file
 ```
3. Created backup:
```yaml
apiVersion: velero.io/v1
kind: Backup
metadata:
  name: mysql-persistent-with-fs
  namespace: openshift-adp
spec:
  includedNamespaces:
  - mysql-persistent
  defaultVolumesToFsBackup: true
```
4. Deleted the `mysql-persistent` namespace and ensured the PV is also removed
5. Performed Restore:
```yaml
apiVersion: velero.io/v1
kind: Restore
metadata:
  name: mysql-persistent
  namespace: openshift-adp
spec:
  backupName:  mysql-persistent-with-fs
  restorePVs: true
```
6. From the mysql pod (restored one)
```shell
$ ls -lah /var/lib/mysql/data/todolist/my_file 
-rwx------. 1 mysql mysql 22 Mar 12 15:41 /var/lib/mysql/data/todolist/my_file
```
7. Checked the node-agent logs for suspicious errors or wornings
8. Ensured node-agent was running as non root, from the node-agent:
```shell
$ id
uid=65534(nobody) gid=65534(nobody) groups=65534(nobody)

$ touch /myfile
touch: cannot touch '/myfile': Read-only file system

# All the files in the /host_pods are mounted as root, example on the restored one
$ ls -lah /host_pods/0b5d3afb-538a-41fe-9021-4de83d460b00/                              
total 24K
drwxr-x---.  5 root root 4.0K Mar 12 15:51 .
drwxr-x---. 53 root root 4.0K Mar 12 16:00 ..
drwxr-x---.  5 root root 4.0K Mar 12 15:51 containers
-rw-r--r--.  1 root root  218 Mar 12 15:51 etc-hosts
drwxr-x---.  3 root root 4.0K Mar 12 15:51 plugins
drwxr-x---.  4 root root 4.0K Mar 12 15:51 volumes
```

9. Node agent spec (FS backup enabled via DPA new spec field)
```yaml
kind: Pod
apiVersion: v1
metadata:
  generateName: node-agent-
  annotations:
    openshift.io/scc: privileged
    seccomp.security.alpha.kubernetes.io/pod: runtime/default
[...]
spec:
[...]
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
  containers:
[...]
      securityContext:
        privileged: true
        readOnlyRootFilesystem: true
        allowPrivilegeEscalation: true
[...]
```

mpryc commented:
/retest ci/prow/unit-test

mpryc commented:
/test unit-test

shubham-pampattiwar commented on internal/controller/nodeagent.go:
```diff
@@ -270,20 +287,42 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
+	}
+
+	// remove HostPods and HostPlugins volumes if not privileged
+	// TODO: Remove this once this is fixed in the velero code
```
remove this TODO ?

mpryc replied:
Updated link to point to 8185.

mateusoliveira43 commented on internal/controller/nodeagent.go:
```diff
@@ -270,20 +287,42 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
 
+	privileged := true
+
+	if dpa.Spec.Configuration.Velero.DisableFsBackup != nil {
```
since you are using default marker, is not this option always set?

mpryc replied:
possibly, but better check then be sorry :)

mateusoliveira43 commented on internal/controller/nodeagent.go:
```diff
@@ -270,20 +287,42 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
-		if vol.Name == HostPlugins {
-			ds.Spec.Template.Spec.Volumes[i].HostPath.Path = getPluginsHostPath(platformType)
+	// update nodeAgent host PV path and plugins host path only if privileged
+	if privileged {
```
you are checking the same condition twice in some parts of the code (for example, in line 299 https://github.com/openshift/oadp-operator/pull/1660/files#diff-080fe4f6649a4b1692368f7a41402aa8c6217c4ad5e5f8afe572fdd87974f939R299). I would rewrite where possible

using here as example 
```go
	if privileged {
		for i, vol := range ds.Spec.Template.Spec.Volumes {
			// update nodeAgent host PV path
			if vol.Name == HostPods && ds.Spec.Template.Spec.Volumes[i].HostPath != nil {
				ds.Spec.Template.Spec.Volumes[i].HostPath.Path = getFsPvHostPath(platformType)
			}
			// update nodeAgent plugins host path
			if vol.Name == HostPlugins && ds.Spec.Template.Spec.Volumes[i].HostPath != nil {
				ds.Spec.Template.Spec.Volumes[i].HostPath.Path = getPluginsHostPath(platformType)
			}
		}
	} else {
		var updatedVolumes []corev1.Volume
		for _, vol := range ds.Spec.Template.Spec.Volumes {
			if vol.Name != HostPods && vol.Name != HostPlugins {
				updatedVolumes = append(updatedVolumes, vol)
			}
		}
		ds.Spec.Template.Spec.Volumes = updatedVolumes
	}
```

mpryc replied:
Refactored function, new commit. Still there are 3 places where the check happens, but now this is more logical:
 1. For the template spec for the volumes
 2. For each of the container volume mounts
 3. For the security context - this could be merged with point 2, but I prefer to keep the code close to the other functions which does set the security context.

shubham-pampattiwar replied:
++

mateusoliveira43 commented on internal/controller/nodeagent.go:
need to also test that 1.4 FSB backups can be restored with new restrictions?

mpryc replied:
Why that could be problematic ? Is the velero backup made by the 1.4 different ?

mateusoliveira43 replied:
one thing that I though reading conversation was if ownership is preserved from 1.4 (root) to current (non root)

mpryc replied:
It is preserved, please check out testing in the comment https://github.com/openshift/oadp-operator/pull/1660#issuecomment-2718388583

Preserved are: ownership , permissions and timestamps.

mpryc commented:
With the last commit:

>>>> Before change
```yaml
kind: Pod
metadata:
  annotations:
    openshift.io/scc: privileged
    [...]
spec:
  [...]
  securityContext:
    runAsUser: 0
  containers:
      securityContext:
        privileged: true
      [...]
      volumeMounts:
        - name: host-pods
          mountPath: /host_pods
          mountPropagation: HostToContainer
        - name: host-plugins
          mountPath: /var/data/kubelet/plugins
          mountPropagation: HostToContainer
        - name: scratch
          mountPath: /scratch
        - name: certs
          mountPath: /etc/ssl/certs
        - name: bound-sa-token
          readOnly: true
          mountPath: /var/run/secrets/openshift/serviceaccount
        - name: kube-api-access-2ht4p
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
```

## Below comparison are against *current* master branch behavior

With:
```yaml
spec:
  configuration:
    velero:
      disableFsBackup: false
```

```yaml
kind: Pod
metadata:
  annotations:
    openshift.io/scc: privileged
    seccomp.security.alpha.kubernetes.io/pod: runtime/default    <---- added
    [...]
spec:
  [...]
  securityContext:
    runAsUser: 0
    runAsNonRoot: false                              <---- added
    seccompProfile:                                  <---- added
      type: RuntimeDefault                           <---- added
  containers:
      securityContext:
        privileged: true
        readOnlyRootFilesystem: true                 <---- added
        allowPrivilegeEscalation: true               <---- added
      [...]
      volumeMounts:
        - name: host-pods
          mountPath: /host_pods
          mountPropagation: HostToContainer
        - name: host-plugins
          mountPath: /var/data/kubelet/plugins
          mountPropagation: HostToContainer
        - name: scratch
          mountPath: /scratch
        - name: certs
          mountPath: /etc/ssl/certs
        - name: bound-sa-token
          readOnly: true
          mountPath: /var/run/secrets/openshift/serviceaccount
        - name: credentials                         <---- added
          mountPath: /tmp/credentials               <---- added
        - name: home-velero                         <---- added
          mountPath: /home/velero                   <---- added
        - name: kube-api-access-8ctns
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
```

With:
```yaml
spec:
  configuration:
    velero:
      disableFsBackup: true
```

```yaml
kind: Pod
metadata:
  annotations:
    openshift.io/scc: restricted-v2                             <---- changed
    seccomp.security.alpha.kubernetes.io/pod: runtime/default   <---- added
    [...]
spec:
  [...]
  securityContext:
    seLinuxOptions:                          <---- added
      level: 's0:c27,c24'                    <---- added
    runAsNonRoot: true                       <---- changed
    fsGroup: 1000750000                      <---- added
    seccompProfile:                          <---- added
      type: RuntimeDefault                   <---- added
  containers:
      securityContext:
        capabilities:                          <---- added
          drop:                                <---- added
            - ALL                              <---- added
        privileged: false                      <---- changed
        runAsUser: 1000750000                  <---- added
        readOnlyRootFilesystem: true           <---- added
        allowPrivilegeEscalation: false        <---- added
      volumeMounts:
        - name: scratch
          mountPath: /scratch
        - name: certs
          mountPath: /etc/ssl/certs
        - name: bound-sa-token
          readOnly: true
          mountPath: /var/run/secrets/openshift/serviceaccount
        - name: credentials                         <---- added
          mountPath: /tmp/credentials               <---- added
        - name: home-velero                         <---- added
          mountPath: /home/velero                   <---- added
        - name: kube-api-access-fgnw2
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
          
     !!!!!!!! Removed:     !!!!!!!! 
        - name: host-pods
          mountPath: /host_pods
          mountPropagation: HostToContainer
        - name: host-plugins
          mountPath: /var/data/kubelet/plugins
          mountPropagation: HostToContainer
```

mpryc commented:
/test 4.19-e2e-test-aws

mpryc commented:
/test 4.19-e2e-test-aws

mpryc commented:
rebased this PR after oadp rebase

mpryc commented:
Do we care about restic backup. Restore for sure? Possibly we need to make /tmp/ also writeable.

```yaml
Velero:   message: /pod volume backup failed: data path backup failed: error running restic backup command restic backup --repo=s3:s3-us-east-1.amazonaws.com/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX/velero-e2e-a439812b-1137-4619-ba0b-1ab5a5c0b6fe/restic/mongo-persistent --password-file=/tmp/credentials/openshift-adp/velero-repo-credentials-repository-password --cache-dir=/scratch/.cache/restic . --tag=pod=mongo-746cc748bf-j4c6d --tag=pod-uid=a9704ab5-2a6e-4af0-952e-32c6c627daa2 --tag=pvc-uid=0643df8a-1b36-40df-af4b-84c8c167e3bf --tag=volume=mongo-data --tag=backup=mongo-restic-e2e-f2ed03dc-03ef-11f0-992f-0a580a80a046 --tag=backup-uid=2c4aa54b-30ee-4186-ab26-a46fb55a0497 --tag=ns=mongo-persistent --host=velero --json with error: exit status 1 stderr: Fatal: unable to save snapshot: open /tmp/restic-temp-pack-678674344: read-only file system
```

mpryc commented:
> I think /tmp should be writable anyway. +1

It is in the last commit.

weshayutin commented:
poking around :)

disableFsBackup: true
```
Velero Container:
bash-5.1$ whoami
1000750000
bash-5.1$ touch /tmp/foo
bash-5.1$ ls /tmp/foo
/tmp/foo
bash-5.1$ touch /var/foo
touch: cannot touch '/var/foo': Permission denied
bash-5.1$ 
```
* basic dm backup passed tests
* fsbackup failed AS EXPECTED, https://termbin.com/hvnd  
   * WE SHOULD DOC the ERROR, so those trying to run fsbackup in a cluster where it's disabled can easily be notified.
*  

Velero
```
oc get pod/velero-8b69fc5f8-wvhns -o yaml | grep scc -A 2 -B 2 
          "dns": {}
      }]
    openshift.io/scc: restricted-v2
    prometheus.io/path: /metrics
    prometheus.io/port: "8085"
```
NonAdmin Controller
```
oc get pod/non-admin-controller-bcd687b88-w2j2q -o yaml | grep scc -A 2 -B 2 
      }]
    openshift.io/oadp-dpa-resource-version: "440051"
    openshift.io/scc: restricted-v2
    seccomp.security.alpha.kubernetes.io/pod: runtime/default
  creationTimestamp: "2025-03-18T19:19:23Z"
```
NodeAgent
```
oc get pod/node-agent-wnt8w -o yaml | grep scc -A 2 -B 2 
          "dns": {}
      }]
    openshift.io/scc: restricted-v2
    seccomp.security.alpha.kubernetes.io/pod: runtime/default
  creationTimestamp: "2025-03-18T19:23:41Z"

```

mpryc commented:
@weshayutin if you are happy we can unhold it and the labels for the oadp ns can be in separate PR as this would be additional ensurance. This way we can hand over this for more detailed testing quicker.

weshayutin commented:
> @weshayutin if you are happy we can unhold it and the labels for the oadp ns can be in separate PR as this would be additional ensurance. This way we can hand over this for more detailed testing quicker.

It's probably better to have the operator adjusting labels in a seperate pr anyway..  +1 from me :)

shubham-pampattiwar commented on internal/controller/nodeagent.go:
```diff
@@ -270,22 +286,57 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
+	}
+
+	ds.Spec.Template.Spec.SecurityContext = &corev1.PodSecurityContext{
+		RunAsNonRoot: ptr.To(!privileged),
```
nice touch !

shubham-pampattiwar commented on internal/controller/nodeagent.go:
```diff
@@ -270,22 +286,57 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
+		// Privileged containers always run as Unconfined seccomp profile
+		// Changing to match the default behavior of the privileged node-agent
+		ds.Spec.Template.Spec.SecurityContext.SeccompProfile = &corev1.SeccompProfile{
+			Type: corev1.SeccompProfileTypeUnconfined,
```
Earlier we were not setting this explicitly, right ?

mpryc replied:
No we were not. We could omit that part, and it would still become Unconfined by default. However, I believe it's better to maintain consistent behavior - if we set SSC for the non-privileged case, we should do the same for this use case as well:

https://kubernetes.io/docs/tutorials/security/seccomp/#before-you-begin

```
Note:
It is not possible to apply a seccomp profile to a container running
with privileged: true set in the container's securityContext.
Privileged containers always run as Unconfined.
```

shubham-pampattiwar commented on internal/controller/nodeagent.go:
```diff
@@ -309,6 +360,35 @@ func (r *DataProtectionApplicationReconciler) customizeNodeAgentDaemonset(ds *ap
 			nodeAgentContainer = &ds.Spec.Template.Spec.Containers[i]
 
+			nodeAgentContainer.SecurityContext = &corev1.SecurityContext{
+				Privileged:               ptr.To(privileged),
```
again, nice touch

mpryc commented:
/unhold
